Feature#12

jcarrier0026 opened the pull request:
I hope I didn't break anything....

jcarrier0026 committed:
Animated Sprite.

Works but is still super messy.

jcarrier0026 committed:
Slightly better code.

It could still use some fine tuning but the animated sprite class works

jcarrier0026 committed:
Forgot Comments.

I forgot to comment on new graphics functions.

mrmarkwell commented on game.cpp:
```diff
@@ -23,14 +23,23 @@ bool Game::Run() {
+
+  // Create an animated sprite object.
+  perf.StartTimer("create_sprite");
+  AnimatedSprite sprite("goku_sprite_sheet(in_progress)",
```
This seems like a strange name for the sprite sheet. Why "in_progress"?

jcarrier0026 replied:
I'm editing the actual .png to make new sprites and adding them to the file. I saved it as (in progress) so we know I haven't finished the final sprite sheet yet.

mrmarkwell commented on game.cpp:
```diff
@@ -23,14 +23,23 @@ bool Game::Run() {
+  perf.StartTimer("create_sprite");
+  AnimatedSprite sprite("goku_sprite_sheet(in_progress)",
+                        {.x = 0, .y = 0, .w = 29, .h = 45},
+                        graphics_.GetRenderer(), 200, {50, 500, 58, 90});
```
For methods with lots of parameters, it can be hard for the reader to know what these all mean (and easy for writer/maintainers to mess something up).

The two ways people deal with this are 1. parameter comments, or 2. declare variables for parameters, or 3. make the parameters a struct.

Examples:

```
// Hard to understand:
foo(10, 25);

// Easier (parameter comments):
foo(/*timer=*/10, /*distance=*/25);

// Also easier (named variable parameters):
constexpr int kTimer = 10;
constexpr int kDistance = 25;
foo(kTimer, kDistance);

// Also easier, define foo to take a struct:

foo({.timer = 10, .distance = 25});

// In this last example, foo would be defined like this:

Struct FooParams {
  int timer = -1;
  int distance = -1;
};
void foo(FooParams params);
```


Here is a tip explaining more:

https://abseil.io/tips/173

mrmarkwell replied:
The main issue here is the "200". I don't know what that is. The destination rectangle could be made a constant too:

`constexpr SDL_Rect kGokuStartLocation = {.x = 50, .y = 500, .w = 58, .h = 90};`

mrmarkwell commented on game.cpp:
```diff
@@ -60,34 +69,15 @@ bool Game::Run() {
     if (input_.WasKeyPressed(SDL_SCANCODE_D) ||
         input_.IsKeyHeld(SDL_SCANCODE_D)) {
-      goku_location += 4;
+      graphics_.PlayAnimation("RunRight", sprite, &perf, false);
```
This is another example like the first comment. I don't know what this 'false' does. I don't want to have to open another file to find out. 

Just add a param comment or make this a variable.

mrmarkwell commented on graphics.cpp:
```diff
@@ -67,3 +46,35 @@ void Graphics::AddSprite(const Sprite& sprite, const SDL_Rect& destination,
+
+void Graphics::AddSprite(AnimatedSprite& sprite, Perf* perf,
+                         std::string animation) {
+  if (sprite.GetVisible()) {
```
For "Getters" that are predicate functions (functions that return true or false), one convention that I like is to name them "IsBlah" instead of "GetBlah", since it makes the code read easier:

```
if (sprite.IsVisible()) {
}
```

mrmarkwell commented on game.cpp:
```diff
@@ -23,14 +23,23 @@ bool Game::Run() {
+  // Create an animated sprite object.
+  perf.StartTimer("create_sprite");
+  AnimatedSprite sprite("goku_sprite_sheet(in_progress)",
+                        {.x = 0, .y = 0, .w = 29, .h = 45},
```
Source location in the sprite sheet is a good candidate to make constant:

`constexpr SDL_Rect kGokuIdleSprite = {.x = 0, .y = 0, .w = 29, .h = 45};`

mrmarkwell commented on graphics.cpp:
```diff
@@ -67,3 +46,35 @@ void Graphics::AddSprite(const Sprite& sprite, const SDL_Rect& destination,
+void Graphics::AddSprite(AnimatedSprite& sprite, Perf* perf,
+                         std::string animation) {
+  if (sprite.GetVisible()) {
+    int frame = sprite.GetFrameIndex();
```
This is odd to make "GetSourceRect" require the frame number to be passed in, when the object that knows the current frame number is the sprite itself. Can't GetSourceRect just take the animation only?

mrmarkwell commented on graphics.cpp:
```diff
@@ -67,3 +46,35 @@ void Graphics::AddSprite(const Sprite& sprite, const SDL_Rect& destination,
+    SDL_Rect source = sprite.GetSourceRect(animation, frame);
+    SDL_Rect destination = sprite.GetDestinationRect();
+    Vector2 offset = sprite.GetOffsets(animation);
+    destination.x = (destination.x + offset.x) % (constants::kWindowWidth - 64);
```
Is Goku always 64 pixels wide? if so, it would be smart to make this 64 another constant in the constants namespace.

mrmarkwell commented on graphics.cpp:
```diff
@@ -67,3 +46,35 @@ void Graphics::AddSprite(const Sprite& sprite, const SDL_Rect& destination,
+  if (sprite.GetVisible()) {
+    int frame = sprite.GetFrameIndex();
+    SDL_Rect source = sprite.GetSourceRect(animation, frame);
+    SDL_Rect destination = sprite.GetDestinationRect();
```
The sprite knows the destination, and the sprite knows the offsets for the given animation, and the sprite needs the new destination rect. This is all AnimatedSprite logic leaking into the graphics class.

Maybe add a new method in animated sprite called "Animate(std::string animation)", and just do the computation inside the sprite class, updating the destination rect internally.

mrmarkwell commented on graphics.cpp:
```diff
@@ -67,3 +46,35 @@ void Graphics::AddSprite(const Sprite& sprite, const SDL_Rect& destination,
+  }
+}
+
+void Graphics::PlayAnimation(std::string animation, AnimatedSprite& sprite,
```
Passing a std::string by value is a no-no since it means you're making a new copy of the string everywhere.

The right way to pass a std::string parameter is either by "const reference" or by using a std::string_view parameter. String_view requires c++17, but it is pretty easy to use:

```
#include <string_view>

// define foo
void foo(std::string_view animation) {}

// use foo
std::string animation = "fart";
foo(animation);
```

mrmarkwell replied:
Although, you really shouldn't be using strings for this. See my other comments about using an enum class.

mrmarkwell commented on animated_sprite.h:
```diff
@@ -0,0 +1,94 @@
+  // the first frame, the renderer, and the time between each frame.
+  AnimatedSprite(std::string sprite_sheet_name,
+                 SDL_Rect first_sprite_coord_location, SDL_Renderer* renderer,
+                 float time_to_update_sprite, SDL_Rect destination);
```
clarify in the comment and in the variable name what units are being use for time.

```
// and the time in milliseconds between each frame.

float update_time_ms
```

mrmarkwell replied:
fix

mrmarkwell commented on animated_sprite.h:
```diff
@@ -0,0 +1,94 @@
+  virtual ~AnimatedSprite() = default;
+  // Checks to see if it is time to move to the next frame in the current
+  // animation.
+  void TimeToUpdateFrame();
```
This method strikes me as odd. It returns nothing and it isn't clear what it does.

Try a better comment and name.

mrmarkwell replied:
fix

mrmarkwell commented on animated_sprite.h:
```diff
@@ -0,0 +1,94 @@
+  // Gets destination_rect_.
+  SDL_Rect GetDestinationRect();
+
+  // Gets frame_index_.
```
nit: These comments are fine but a little technical. Consider writing them like the person reading doesn't know anything about the internals of the class.

```
// Returns the index to the current frame being displayed for this animated sprite.
int GetFrameIndex();
```

mrmarkwell commented on animated_sprite.h:
```diff
@@ -0,0 +1,94 @@
+
+ private:
+  // Holds animation name and a vector with all frames for an animation.
+  std::unordered_map<std::string, std::vector<SDL_Rect>> animations_;
```
I don't like that the animations are named with std::strings and not with an enum class, but don't worry about it for this change. I can refactor this later.

An enum would give us very clear types:

```
enum class Animation {
  kIdle,
  kRun,
  kPunch,
};
```

Using std::strings makes it like the wild west - anything goes. Really easy to make a mistake.

Like I said, don't worry about it for now unless you want to.

mrmarkwell commented on animated_sprite.h:
```diff
@@ -0,0 +1,94 @@
+  std::unordered_map<std::string, std::vector<SDL_Rect>> animations_;
+  // Holds the x and y offsets for each animation.
+  std::unordered_map<std::string, Vector2> offsets_;
+  // Tells you what frame of an animation you are on.
```
Give these primitives (ints and bools) a default. They are not initialized with a default value so leaving them uninitialized can be dangerous (if you forgot to initialize them elsewhere).

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,113 @@
+      destination_(destination) {
+  this->SetupAnimations();
+}
+
```
this-> is not necessary inside methods in the class. Just "SetupAnimations" is fine, the this-> is implied.

mrmarkwell replied:
Please fix

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,113 @@
+      frame_index_++;
+    } else {
+      if (current_animation_once == true) {
+        this->SetVisible(false);
```
remove "this->"

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,113 @@
+  this->SetupAnimations();
+}
+
+// TimeToUpdateFrame()
```
Remove this comment.

Function comments should be in the header file, not the .cc file. The header file is the documentation of the class.

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,113 @@
+  }
+}
+
+// GetCurrentAnimation()
```
Remove all these comments

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,113 @@
+// TimeToUpdateFrame()
+void AnimatedSprite::TimeToUpdateFrame() {
+  if (elapsed_time_ >= time_to_update) {
+    if (frame_index_ < (int)animations_[this->current_animation].size() - 1) {
```
Remove this-> everywhere in this file.

Prefer c++ casts over c-style casts:

`static_cast<int>(thing);`

They are safer and more explicit:
https://stackoverflow.com/questions/1609163/what-is-the-difference-between-static-cast-and-c-style-casting/18414126#:~:text=C%2B%2B%20style%20casts%20are%20checked,of%20the%20programmer%20more%20clearly.

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,113 @@
+    if (frame_index_ < (int)animations_[this->current_animation].size() - 1) {
+      frame_index_++;
+    } else {
+      if (current_animation_once == true) {
```
The comparison to true isn't necessary:

`if (current_animation_once)`

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,113 @@
+void AnimatedSprite::TimeToUpdateFrame() {
+  if (elapsed_time_ >= time_to_update) {
+    if (frame_index_ < (int)animations_[this->current_animation].size() - 1) {
+      frame_index_++;
```
You just want to loop from 0 to animations.size. You can do this with the mod operator:
`frame_index = frame_index % animations_[current_animation].size();`

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,113 @@
+
+// SetDestinationRect()
+void AnimatedSprite::SetDestinationRect(SDL_Rect new_rect) {
+  destination_.x = new_rect.x;
```
destination_ = new_rect;

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,113 @@
+}
+
+// SetDestinationRect()
+void AnimatedSprite::SetDestinationRect(SDL_Rect new_rect) {
```
Take these parameters by const ref unless you can't. Passing by value makes a copy.

```
foo(const SDL_Rect& new_destination) {}
```

mrmarkwell replied:
fix

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,113 @@
+                                  SDL_Rect first_sprite_coord_location,
+                                  Vector2 offset) {
+  std::vector<SDL_Rect> rectangles;
+  SDL_Rect new_rect;
```
nit: Define new_rect where it is first used (two lines down).

mrmarkwell replied:
fix

mrmarkwell replied:
This is marked resolved but I don't see that it is fixed.

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,113 @@
+    new_rect.x += (i * first_sprite_coord_location.w) + 1;
+    rectangles.push_back(new_rect);
+  }
+  animations_.insert(
```
Having two maps with the same key is a code-smell. If it were me, I'd make a struct called Animation like so:

```
enum class AnimationType {
   kIdle,
   kRun,
};

struct Animation {
  int frames;
  Vector2 offset;
  SDL_rect start_rect; // The other rects can be computed when needed.
};
```

Then you can just have a single map from animation type to animation struct:

```
// Holds animation name and a vector with all frames for an animation.
  std::unordered_map<AnimationType, Animation> animations_;
```

And you can just get rid of the "offsets_" map completely.

You don't need a vector of rectangles either, since computing the rectangle based on the starting rectangle is easy.

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,113 @@
+void AnimatedSprite::SetVisible(bool visible) { visible_ = visible; }
+
+// AddAnimation()
+void AnimatedSprite::AddAnimation(std::string name, int frames,
```
Actually I think you should just get rid of this method, and make animations be passed in with the constructor. See my comment on this below.

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,113 @@
+  offsets_.clear();
+}
+
+// SetupAnimations()
```
This isn't right: this is hardcoding goku's animations.

Make this take a std::unordered_map, and then define the animation map as a constexpr where you create the AnimatedSprite. This should probably be a parameter for the constructor.

```
constexpr std::unordered_map<AnimationType, Animation> kGokuAnimations =
{
  {AnimationType::kIdle, {.frames = 2, .offset = {.x = 0, .y = 0}, .start_rect = {0, 0, 29, 45}}}
  {AnimationType::kRunRight, /*other stuff...*/
}
```

clang-format will make this look nice for you.

mrmarkwell commented on animated_sprite.h:
```diff
@@ -0,0 +1,94 @@
+
+  // Gets offsets.
+  Vector2 GetOffsets(std::string animation);
+
```
Having tons of getters and setters is a code-smell. The AnimatedSprite class should encapsulate as much as possible.

If you have getters and setters for everything, this basically just becomes a struct.

mrmarkwell commented on graphics.h:
```diff
@@ -30,6 +21,15 @@ class Graphics {
 
+  // Adds the current frame in the animation to the renderer. It also adds the
+  // offsets for the destination rectangle.
+  void AddSprite(AnimatedSprite& sprite, Perf* perf, std::string animation);
```
Good job passing the AnimatedSprite by reference. Can it be by const reference?

mrmarkwell commented on sprite.cpp:
```diff
@@ -10,22 +10,22 @@ constexpr char kAssetPath[] = "assets/";
-        SDL_Texture* texture = SDL_CreateTextureFromSurface(renderer, surface);
-        sprite_textures_[sprite_sheet_name] = texture;
-      }
+  // Check if we've already opened this sprite sheet.
```
Looks like there are some whitespace changes here. Can you run clang-format on all the files to make sure this is right?

mrmarkwell reviewed (changes requested):
Overall, this is really great. I made a lot of suggestions but I think the general additions and logic here is fantastic. Great work!

I'm down to chat about any of these suggestions, just text me.

jcarrier0026 committed:
Updated AnimatedSprite.

Made changes to the AnimatedSprite class to make it more comprehensible.

jcarrier0026 committed:
I'm stupid.

Here is animation.h.....

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,142 @@
+  if (elapsed_time_ >= animations_[current_animation].time_between_frames) {
+    elapsed_time_ = 0;
+    frame_index_ =
+        (frame_index_ + 1) % (int)animations_[current_animation].num_of_frames;
```
Use c++ casts 

static_cast<int>

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,142 @@
+  }
+}
+
+SDL_Rect AnimatedSprite::GetDestinationRect() {
```
This seems wrong - you're adding the offset to the current destination rectangle. What if this is called a bunch of times? It seems like it would move the destination rectangle around on the screen.

jcarrier0026 replied:
TODO: Needs to be changed.

mrmarkwell replied:
SGTM

jcarrier0026 committed:
Priority Animations

Updated the code to have certain animations prioritized over others.

jcarrier0026 commented on animated_sprite.cpp:
```diff
@@ -18,15 +18,13 @@ AnimatedSprite::AnimatedSprite(std::string sprite_sheet_name,
 
-void AnimatedSprite::TimeToUpdateFrame() {
+bool AnimatedSprite::TimeToUpdateFrame() {
+  // Here
```
Need to delete this

mrmarkwell commented on animated_sprite.h:
```diff
@@ -0,0 +1,72 @@
+  // Keeps track of which frame of the animation is currently on screen.
+  int frame_index_;
+  // Tells you how much time has passed since the start of a frame.
+  int elapsed_time_ = 0;
```
What units?

mrmarkwell commented on animated_sprite.h:
```diff
@@ -0,0 +1,72 @@
+  // Tells you how much time has passed since the start of a frame.
+  int elapsed_time_ = 0;
+  // Tells you when the current frame started.
+  int last_time_ = 0;
```
units? Will this overflow?

mrmarkwell commented on animated_sprite.h:
```diff
@@ -0,0 +1,72 @@
+  // Where the sprite is going to be print on the screen.
+  SDL_Rect destination_;
+  // Used to setup all animations for the sprite.
+  Animation animation;
```
What is the difference between this animation and the ones in the map?

mrmarkwell replied:
This isn't needed anymore.

mrmarkwell commented on animated_sprite.h:
```diff
@@ -0,0 +1,72 @@
+  void AddAnimation();
+
+  // Clears the animations_ map.
+  void ResetAnimation();
```
This probably isn't needed

mrmarkwell commented on animation.h:
```diff
@@ -0,0 +1,21 @@
+
+struct Animation {
+  AnimationType name;
+  int num_of_frames;
```
This is probably just image_rects.size(), right?

mrmarkwell commented on animation.h:
```diff
@@ -0,0 +1,21 @@
+  SDL_Renderer *renderer;
+  Vector2 offsets;
+
+  Animation &operator=(Animation new_animation);
```
Let's talk about this

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,142 @@
+  int num_of_frames = 2;
+  bool once = false;
+
+  animation = {AnimationType::kIdle,
```
I see what you're doing. This is clever, but there is an easier way to do this. Let's chat about it.

mrmarkwell replied:
```
AddAnimation({AnimationType::kIdle,
               num_of_frames,
               source_rectangle,
               time_between_frames,
               rect,
               animation.sprite_sheet_name,
               once,
               animation.renderer,
               Vector2(0, 0),
               priority});
```

mrmarkwell commented on game.cpp:
```diff
@@ -23,14 +23,30 @@ bool Game::Run() {
+  Sprite background("Namek Background", source_rect, graphics_.GetRenderer());
+
+  // Create an animated sprite object.
+  // DELETE.
```
What's with the DELETE's here?

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,158 @@
+}
+
+SDL_Rect AnimatedSprite::GetDestinationRect() {
+  destination_.x = (destination_.x + animations_[current_animation].offsets.x) %
```
Make a new issue to separate sprite location on the screen from the concept of the animation.

mrmarkwell commented on animated_sprite.h:
```diff
@@ -0,0 +1,75 @@
+
+ private:
+  // Keeps track of what animation is playing.
+  AnimationType current_animation;
```
current_animation_;

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,158 @@
+}
+
+void AnimatedSprite::ChangeAnimation(AnimationType animation, bool once) {
+  animations_[animation].once = once;
```
Any mutable data should be moved out of the Animation struct since it isn't really part of the definition of that animation.

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,158 @@
+      animations_[current_animation].priority) {
+    ChangeAnimation(animation, once);
+  } else if (TimeToUpdateFrame()) {
+    if (current_animation != animation &&
```
Comment please

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,158 @@
+  }
+}
+
+Animation& Animation::operator=(Animation new_animation) {
```
This is not needed, but very cool.

mrmarkwell replied:
Compiler autogenerates:

1. Ctor
2. Dtor
3. Copy ctor
4. Copy assignment operator (that's what you wrote here)
5. Move ctor
6. Move assignment operator

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,158 @@
+  return *this;
+}
+
+void AnimatedSprite::ResetAnimation() { animations_.clear(); }
```
Remove, we won't need this

mrmarkwell commented on animation.h:
```diff
@@ -0,0 +1,22 @@
+  int time_between_frames;
+  std::vector<SDL_Rect> image_rects;
+  std::string sprite_sheet_name;
+  bool once;
```
Move this out since it changes

mrmarkwell commented on game.cpp:
```diff
@@ -59,35 +75,28 @@ bool Game::Run() {
+      play_animation_once = false;
+      sprite.PlayAnimation(AnimationType::kRunLeft, &perf, play_animation_once);
+    }
+    // DELETE.
```
// TODO(markwell): Move these key presses into the Player class.

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -0,0 +1,158 @@
+  return destination_;
+}
+
+SDL_Rect AnimatedSprite::GetCurrentFrameRect() {
```
```
SDL_Rect AnimatedSprite::GetCurrentFrameRect() const {
  return animations_[current_animation].image_rects[frame_index_];
}
```

jcarrier0026 replied:
TODO: Couldn't get this to work. We can try to fix this together.

mrmarkwell replied:
SGTM

jcarrier0026 committed:
Fixed AnimatedSprite.

I resolved all requested changes except for one. We can look at it together later.

mrmarkwell reviewed (approved):
Looks great, ship it

jcarrier0026 commented on animated_sprite.cpp:
```diff
@@ -4,53 +4,50 @@
   SDL_Rect new_rect;
-  for (int i = 0; i < animation.num_of_frames; i++) {
+  for (int i = 0; i < num_of_frames; i++) {
+    // new_rect is updated with the source rectangle of the next frame of the
```
This is where I put the comment for the issue above @mrmarkwell
